nullaway/src/test/java/com/uber/nullaway/SyncLambdasTests.java

```diff
@@ -0,0 +1,163 @@
+            "import org.jspecify.annotations.Nullable;",
+            "public class Test {",
+            "    private @Nullable HashMap<Object, Object> target;",
+            "    private @Nullable HashMap<Object, Object> resolved;",
+            "    public void initialize() {",
+            "        if (this.target == null) {",
+            "            throw new IllegalArgumentException();",
+            "        }",
+            "        this.resolved = new HashMap<>();",
+            "        this.target.forEach((key, value) -> {",
+            "            // no error here as info gets propagated",
+            "            this.resolved.put(key, value);",
+            "        });",
+            "    }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void otherForEach() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import java.util.HashMap;",
+            "import java.util.function.BiConsumer;",
+            "import org.jspecify.annotations.Nullable;",
+            "public class Test {",
+            "    private @Nullable MyMap<Object, Object> target;",
+            "    private @Nullable MyMap<Object, Object> resolved;",
+            "    static class MyMap<K,V> {",
+            "        public void forEach(BiConsumer<Object, Object> consumer) {}",
+            "        public void put(Object key, Object value) {}",
+            "    }",
+            "    public void initialize() {",
+            "        if (this.target == null) {",
+            "            throw new IllegalArgumentException();",
+            "        }",
+            "        this.resolved = new MyMap<>();",
+            "        this.target.forEach((key, value) -> {",
+            "            // error since this is a custom type, not inheriting from java.util.Map",
+            "            // BUG: Diagnostic contains: dereferenced expression this.resolved is @Nullable",
+            "            this.resolved.put(key, value);",
+            "        });",
+            "    }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void forEachOnIterable() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import java.util.List;",
+            "import java.util.ArrayList;",
+            "import org.jspecify.annotations.Nullable;",
+            "public class Test {",
+            "    private @Nullable Object f;",
+            "    public void test1() {",
+            "        if (this.f == null) {",
+            "            throw new IllegalArgumentException();",
+            "        }",
+            "        List<Object> l = new ArrayList<>();",
+            "        l.forEach(v -> System.out.println(v + this.f.toString()));",
+            "        Iterable<Object> l2 = l;",
+            "        l2.forEach(v -> System.out.println(v + this.f.toString()));",
```

Comment:
Worth adding a case where we set `this.f = null` then do one of these and show it now reports an error?

Reply:
26bc3a06959bdcc6c3eaa012f8e9b761576b3c48